src/IW/Sync/Search.hs

```
@@ -1,97 +1,214 @@
+       | otherwise     -> githubSearchAll paths properties recent (pred interval) acc
+  where
+    -- | The first day of the search interval. It's calculated by subtracting the size of the
+    -- interval from the most recent day of the search interval.
+    firstDay :: Day
+    firstDay = negate interval `addDays` recent
 
--- | Search all repositories built with the Haskell language.
-searchHaskellRepos :: IO (Either GitHub.Error (SearchResult GitHub.Repo))
-searchHaskellRepos = searchRepos "language:haskell"
+    searchResultToList :: SearchResult a -> [a]
+    searchResultToList (SearchResult _ vec) = V.toList vec
 
--- | Construct a github search query from a list of labels.
-labelsToSearchQuery :: [Label] -> Text
-labelsToSearchQuery = foldMap (\Label{..} -> "label:\"" <> unLabel <> "\" ")
+-- | Executes a GitHub search action within the context of the @App@ monad.
+executeGithubSearch
+    :: forall a env m.
+       ( MonadIO m
+       , MonadUnliftIO m
+       , WithError m
+       , WithLog env m
+       , FromJSON a
+       )
+    => Paths -- ^ Query paths
+    -> Text  -- ^ Query properties
+    -> Day   -- ^ The first day of the date interval
+    -> Day   -- ^ The last day of the date interval
+    -> Int   -- ^ The number of the page to be searched
+    -> m (SearchResult a)
+executeGithubSearch paths properties from to page = do
+    searchLimit <- getSearchRateLimit
+    log I $ "Current search limit information: " <> show searchLimit
+    if limitsRemaining searchLimit > 0
+        then do
+            log D $ "Searching GitHub API with following request: " <> showGithubQuery paths properties from to page
+            liftIO (githubSearch paths properties from to page) >>= \case
+                Left err -> throwError $ githubErrToAppErr err
+                Right searchRes -> pure searchRes
+        else do
+            log I "API limit reached. Delaying search..."
+            threadDelay 60000000
```

Let's just throw an error here. The limits for GitHub API is 5000 per hour. So waiting one minute doesn't really help... If we hit the limit, we will handle it separately. You can create the issue to handle cases when we reach rate limit.